src/components/text-editor/prosemirror-adapter/plugins/trigger-popup-plugin.ts

```diff
@@ -0,0 +1,125 @@
+import { Plugin, PluginKey, EditorState } from 'prosemirror-state';
+import { ReplaceStep, ReplaceAroundStep } from 'prosemirror-transform';
+import { EditorView } from 'prosemirror-view';
+
+export const triggerPluginKey = new PluginKey('triggerPlugin');
+
+type Trigger = '@' | '#';
+
+const triggers: Trigger[] = ['@', '#'];
+
+const isTrigger = (key: string): key is Trigger => {
+    return triggers.includes(key as Trigger);
+};
+
+const shouldTrigger = (state: EditorState): boolean => {
+    const { $from, $to } = state.selection;
+
+    const prevPos = $from.pos - 1;
+
+    if (!prevPos) {
```

Comment:
I think we generally try to avoid abbreviations. So maybe reconsider `pos` or `char` or `prev` and such 😊